script.keymap/resources/lib/editor.py

```diff
@@ -0,0 +1,129 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2014 Thomas Amland
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from kodi_six import xbmc, xbmcaddon
+from threading import Timer
+from collections import OrderedDict
+from kodi_six.xbmcgui import Dialog, WindowXMLDialog
+from resources.lib.actions import ACTIONS, WINDOWS
+from resources.lib.utils import tr
+
+KODIMONITOR = xbmc.Monitor()
+
+class Editor(object):
+    def __init__(self, defaultkeymap, userkeymap):
+        """Create the editor object."""
+        self.defaultkeymap = defaultkeymap
+        self.userkeymap = userkeymap
+        self.dirty = False
+
+    def start(self):
+        while not KODIMONITOR.abortRequested():
+            # Select context menu
+            idx = Dialog().select(tr(30007), list(WINDOWS.values()))
+            if idx == -1:
+                break
+            window = list(WINDOWS.keys())[idx]
+
+            while not KODIMONITOR.abortRequested():
+                # Select category menu
+                idx = Dialog().select(tr(30008), list(ACTIONS.keys()))
+                if idx == -1:
+                    break
+                category = list(ACTIONS.keys())[idx]
+
+                while True:
```

Comment:
forever running loop. Note that there isn't any pause on the loop which will for sure consume too much CPU while this method is in use

Reply:
I've added an abort request check to the loop.  I looked through the code again, and the select dialog boxes act as pauses in the loops, so there don't appear to be any race conditions here. I also tested by bringing up a dialog box and then quitting Kodi.  Kodi quit within the normal amount of time and the log didn't show any forced terminations of python threads.